File: dotnet/src/LlmInferenceProvider.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

using GitHub.Copilot.Rpc;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Channels;

namespace GitHub.Copilot;

/// <summary>
/// Transport the runtime would otherwise use to issue an intercepted
/// model-layer request.
/// </summary>
[Experimental(Diagnostics.Experimental)]
public enum LlmInferenceTransport
{
    /// <summary>
    /// Plain HTTP or a streamed SSE response. Each body chunk is an opaque
    /// byte range.
    /// </summary>
    Http,

    /// <summary>
    /// Full-duplex WebSocket channel. Each request-body chunk is one inbound
    /// WebSocket message and each response-body write is one outbound message.
    /// </summary>
    WebSocket,
}

/// <summary>
/// An outbound model-layer HTTP (or WebSocket) request the runtime is asking
/// the SDK consumer to service on its behalf.
/// </summary>
/// <remarks>
/// This is a low-level shape: URL / method / headers verbatim, body bytes
/// delivered as an async sequence, and the response delivered through the
/// <see cref="ResponseBody"/> sink. The runtime does not classify the request
/// (no provider type, endpoint kind, or wire API); consumers that need that
/// information derive it from the URL / headers themselves.
/// </remarks>
[Experimental(Diagnostics.Experimental)]
public sealed class LlmInferenceRequest
{
    /// <summary>Opaque runtime-minted id, stable across the request lifecycle.</summary>
    public required string RequestId { get; init; }

    /// <summary>
    /// Id of the runtime session that triggered this request, when one is in
    /// scope. <see langword="null"/> for out-of-session requests (e.g. startup
    /// model catalog).
    /// </summary>
    public string? SessionId { get; init; }

    /// <summary>HTTP method (<c>GET</c>, <c>POST</c>, ...).</summary>
    public required string Method { get; init; }

    /// <summary>Absolute request URL.</summary>
    public required string Url { get; init; }

    /// <summary>HTTP request headers, lowercased names mapped to multi-valued lists.</summary>
    public required IReadOnlyDictionary<string, IReadOnlyList<string>> Headers { get; init; }

    /// <summary>
    /// Transport the runtime would otherwise use. <see cref="LlmInferenceTransport.Http"/>
    /// covers plain HTTP and SSE responses; <see cref="LlmInferenceTransport.WebSocket"/>
    /// indicates a full-duplex message channel. Consumers branch on this to
    /// decide whether to service the request with an HTTP client or a WebSocket
    /// client.
    /// </summary>
    public LlmInferenceTransport Transport { get; init; }

    /// <summary>
    /// Request body bytes, yielded as they arrive from the runtime. Always
    /// enumerable; an empty body yields zero chunks before completing. For
    /// WebSocket transport each element is one inbound message.
    /// </summary>
    public required IAsyncEnumerable<ReadOnlyMemory<byte>> RequestBody { get; init; }

    /// <summary>
    /// Cancelled when the runtime aborts this in-flight request (e.g. the agent
    /// turn was aborted upstream). Pass it straight to <c>HttpClient.SendAsync</c>
    /// / your transport so the upstream call is torn down too. After it fires,
    /// writes to <see cref="ResponseBody"/> are ignored.
    /// </summary>
    public CancellationToken CancellationToken { get; init; }

    /// <summary>
    /// Sink the consumer writes the upstream response into. Call
    /// <see cref="LlmInferenceResponseSink.StartAsync"/> exactly once before
    /// writing body chunks, then zero or more
    /// <see cref="LlmInferenceResponseSink.WriteAsync(ReadOnlyMemory{byte})"/>
    /// calls, and finish with <see cref="LlmInferenceResponseSink.EndAsync"/> or
    /// <see cref="LlmInferenceResponseSink.ErrorAsync"/>.
    /// </summary>
    public required LlmInferenceResponseSink ResponseBody { get; init; }
}

/// <summary>Response head passed to <see cref="LlmInferenceResponseSink.StartAsync"/>.</summary>
[Experimental(Diagnostics.Experimental)]
public sealed class LlmInferenceResponseInit
{
    /// <summary>HTTP status code (101 acknowledges a WebSocket upgrade).</summary>
    public int Status { get; init; }

    /// <summary>Optional HTTP status reason phrase.</summary>
    public string? StatusText { get; init; }

    /// <summary>Response headers, lowercased names mapped to multi-valued lists.</summary>
    public IReadOnlyDictionary<string, IReadOnlyList<string>>? Headers { get; init; }
}

/// <summary>
/// Sink the consumer writes the upstream response into. The state machine is
/// strict: <see cref="StartAsync"/> once → zero or more <c>WriteAsync</c> →
/// exactly one of <see cref="EndAsync"/> or <see cref="ErrorAsync"/>. Calling
/// out of order throws.
/// </summary>
[Experimental(Diagnostics.Experimental)]
public abstract class LlmInferenceResponseSink
{
    /// <summary>Sends the response head (status + headers) back to the runtime.</summary>
    public abstract Task StartAsync(LlmInferenceResponseInit init);

    /// <summary>Sends a binary body chunk (base64-encoded on the wire).</summary>
    public abstract Task WriteAsync(ReadOnlyMemory<byte> data);

    /// <summary>Sends a UTF-8 text body chunk.</summary>
    public abstract Task WriteAsync(string text);

    /// <summary>Marks end-of-stream cleanly.</summary>
    public abstract Task EndAsync();

    /// <summary>Marks end-of-stream with a transport-level failure.</summary>
    public abstract Task ErrorAsync(string message, string? code = null);
}

/// <summary>
/// Implemented by SDK consumers to service the LLM inference requests the
/// runtime would otherwise issue itself. The same callback handles both
/// buffered and streaming responses — the consumer just calls
/// <see cref="LlmInferenceResponseSink.WriteAsync(ReadOnlyMemory{byte})"/> zero
/// or more times before <see cref="LlmInferenceResponseSink.EndAsync"/>.
/// </summary>
/// <remarks>
/// Prefer subclassing <see cref="LlmRequestHandler"/> for a transparent
/// pass-through starting point; implement this interface directly only when you
/// need full control over the raw byte streams.
/// </remarks>
[Experimental(Diagnostics.Experimental)]
public interface ILlmInferenceProvider
{
    /// <summary>
    /// Invoked by the runtime once per outbound LLM request the consumer has
    /// opted to handle. The consumer is responsible for eventually calling
    /// either <see cref="LlmInferenceResponseSink.EndAsync"/> or
    /// <see cref="LlmInferenceResponseSink.ErrorAsync"/>; failing to do so leaks
    /// runtime state. Throwing surfaces a transport-level failure to the runtime
    /// (equivalent to <c>ResponseBody.ErrorAsync(...)</c> when
    /// <see cref="LlmInferenceResponseSink.StartAsync"/> has not yet been called).
    /// </summary>
    Task OnLlmRequestAsync(LlmInferenceRequest request);
}

/// <summary>
/// Adapts an <see cref="ILlmInferenceProvider"/> into the generated
/// <see cref="ILlmInferenceHandler"/> shape consumed by the SDK's RPC
/// dispatcher.
/// </summary>
/// <remarks>
/// Maintains a per-<c>requestId</c> state table: each <c>httpRequestStart</c>
/// allocates a body channel + response sink and fires
/// <see cref="ILlmInferenceProvider.OnLlmRequestAsync"/> in the background.
/// Subsequent <c>httpRequestChunk</c> frames are routed into the channel. The
/// sink translates <c>Start</c> / <c>Write</c> / <c>End</c> / <c>Error</c> calls
/// into outbound <c>llmInference.httpResponseStart</c> /
/// <c>llmInference.httpResponseChunk</c> calls.
/// </remarks>
internal sealed class LlmInferenceAdapter : ILlmInferenceHandler
{
    private readonly ILlmInferenceProvider _provider;
    private readonly Func<ILlmInferenceResponseChannel?> _getChannel;
    private readonly ConcurrentDictionary<string, PendingState> _pending = new(StringComparer.Ordinal);

    // Defense-in-depth backstop: chunks that arrive before their start frame
    // (a reordering the runtime's single ordered dispatch should make
    // impossible) are staged here and drained the moment httpRequestStart
    // registers the matching state, so a body byte is never silently dropped.
    private readonly ConcurrentDictionary<string, List<LlmInferenceHttpRequestChunkRequest>> _staged = new(StringComparer.Ordinal);

    internal LlmInferenceAdapter(ILlmInferenceProvider provider, Func<ServerRpc?> getServerRpc)
        : this(provider, WrapServerRpc(getServerRpc ?? throw new ArgumentNullException(nameof(getServerRpc))))
    {
    }

    internal LlmInferenceAdapter(ILlmInferenceProvider provider, Func<ILlmInferenceResponseChannel?> getChannel)
    {
        _provider = provider ?? throw new ArgumentNullException(nameof(provider));
        _getChannel = getChannel ?? throw new ArgumentNullException(nameof(getChannel));
    }

    /// <summary>
    /// Adapts a <see cref="ServerRpc"/> getter into a response-channel getter,
    /// caching the wrapper so a new one is allocated only when the underlying
    /// connection changes (e.g. reconnect).
    /// </summary>
    private static Func<ILlmInferenceResponseChannel?> WrapServerRpc(Func<ServerRpc?> getServerRpc)
    {
        ServerRpc? cachedRpc = null;
        ILlmInferenceResponseChannel? cachedChannel = null;
        return () =>
        {
            var rpc = getServerRpc();
            if (rpc is null)
            {
                return null;
            }

            if (!ReferenceEquals(rpc, cachedRpc))
            {
                cachedRpc = rpc;
                cachedChannel = new ServerRpcResponseChannel(rpc);
            }

            return cachedChannel;
        };
    }

    public Task<LlmInferenceHttpRequestStartResult> HttpRequestStartAsync(LlmInferenceHttpRequestStartRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var state = new PendingState();
        _pending[request.RequestId] = state;

        if (_staged.TryRemove(request.RequestId, out var stagedChunks))
        {
            foreach (var chunk in stagedChunks)
            {
                RouteChunk(state, chunk);
            }
        }

        var sink = new AdapterResponseSink(request.RequestId, state, _getChannel, _pending);
        state.Sink = sink;

        var transport = request.Transport == LlmInferenceHttpRequestStartTransport.Websocket
            ? LlmInferenceTransport.WebSocket
            : LlmInferenceTransport.Http;

        var llmRequest = new LlmInferenceRequest
        {
            RequestId = request.RequestId,
            SessionId = request.SessionId,
            Method = request.Method,
            Url = request.Url,
            Headers = ToReadOnlyHeaders(request.Headers),
            Transport = transport,
            RequestBody = state.Body.ReadAllAsync(state.Abort.Token),
            CancellationToken = state.Abort.Token,
            ResponseBody = sink,
        };

        // Return from httpRequestStart immediately (after registering state) so
        // the runtime's RPC reply is not gated on the consumer's I/O. The actual
        // provider work runs asynchronously.
        _ = RunProviderAsync(llmRequest, state, sink);

        return Task.FromResult(new LlmInferenceHttpRequestStartResult());
    }

    public Task<LlmInferenceHttpRequestChunkResult> HttpRequestChunkAsync(LlmInferenceHttpRequestChunkRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        if (_pending.TryGetValue(request.RequestId, out var state))
        {
            RouteChunk(state, request);
        }
        else
        {
            _staged.AddOrUpdate(
                request.RequestId,
                _ => [request],
                (_, list) =>
                {
                    list.Add(request);
                    return list;
                });
        }

        return Task.FromResult(new LlmInferenceHttpRequestChunkResult());
    }

    private async Task RunProviderAsync(LlmInferenceRequest request, PendingState state, AdapterResponseSink sink)
    {
        try
        {
            await _provider.OnLlmRequestAsync(request).ConfigureAwait(false);
            if (!state.Finished)
            {
                await FailViaSink(
                    sink,
                    state,
                    "LLM inference provider returned without finalising the response (call ResponseBody.EndAsync() or .ErrorAsync()).").ConfigureAwait(false);
            }
        }
        catch (Exception ex)
        {
            if (state.Cancelled || state.Abort.IsCancellationRequested)
            {
                // The runtime already cancelled this request; the provider's
                // throw is just the abort propagating out of its upstream call.
                await FinishCancelled(sink, state).ConfigureAwait(false);
                return;
            }

            await FailViaSink(sink, state, ex.Message).ConfigureAwait(false);
        }
    }

    private static async Task FailViaSink(AdapterResponseSink sink, PendingState state, string message)
    {
        if (state.Finished)
        {
            return;
        }

        try
        {
            if (!state.Started)
            {
                await sink.StartAsync(new LlmInferenceResponseInit { Status = 502 }).ConfigureAwait(false);
            }

            await sink.ErrorAsync(message).ConfigureAwait(false);
        }
        catch
        {
            // Best-effort — the connection may already be dead.
        }
    }

    private static async Task FinishCancelled(AdapterResponseSink sink, PendingState state)
    {
        if (state.Finished)
        {
            return;
        }

        try
        {
            if (!state.Started)
            {
                await sink.StartAsync(new LlmInferenceResponseInit { Status = 499 }).ConfigureAwait(false);
            }

            await sink.ErrorAsync("Request cancelled by runtime", "cancelled").ConfigureAwait(false);
        }
        catch
        {
            // Best-effort — the runtime already dropped the request on cancel.
        }
    }

    private static void RouteChunk(PendingState state, LlmInferenceHttpRequestChunkRequest chunk)
    {
        if (chunk.Cancel == true)
        {
            state.Cancelled = true;
            state.Abort.Cancel();
            state.Body.PushCancel(chunk.CancelReason);
            return;
        }

        if (!string.IsNullOrEmpty(chunk.Data))
        {
            state.Body.PushChunk(DecodeChunkData(chunk.Data, chunk.Binary == true));
        }

        if (chunk.End == true)
        {
            state.Body.PushEnd();
        }
    }

    private static byte[] DecodeChunkData(string data, bool binary) =>
        binary ? Convert.FromBase64String(data) : Encoding.UTF8.GetBytes(data);

    private static Dictionary<string, IReadOnlyList<string>> ToReadOnlyHeaders(IDictionary<string, IList<string>> headers)
    {
        var result = new Dictionary<string, IReadOnlyList<string>>(StringComparer.OrdinalIgnoreCase);
        foreach (var (name, values) in headers)
        {
            result[name] = values as IReadOnlyList<string> ?? [.. values];
        }

        return result;
    }

    private sealed class PendingState
    {
        public BodyChannel Body { get; } = new();

        public CancellationTokenSource Abort { get; } = new();

        public bool Started { get; set; }

        public bool Finished { get; set; }

        public bool Cancelled { get; set; }

        public AdapterResponseSink? Sink { get; set; }
    }

    /// <summary>
    /// An unbounded channel of request-body items exposed as an
    /// <see cref="IAsyncEnumerable{T}"/> of byte chunks. A cancel item surfaces
    /// as an <see cref="OperationCanceledException"/> out of the enumerator so
    /// the consumer's upstream call is torn down.
    /// </summary>
    private sealed class BodyChannel
    {
        private readonly Channel<Item> _channel = Channel.CreateUnbounded<Item>(
            new UnboundedChannelOptions { SingleReader = true, SingleWriter = true });

        public void PushChunk(byte[] data) => _channel.Writer.TryWrite(new Item { Chunk = data });

        public void PushEnd() => _channel.Writer.TryWrite(new Item { End = true });

        public void PushCancel(string? reason) => _channel.Writer.TryWrite(new Item { Cancel = true, CancelReason = reason });

        public async IAsyncEnumerable<ReadOnlyMemory<byte>> ReadAllAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
        {
            while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
            {
                while (_channel.Reader.TryRead(out var item))
                {
                    if (item.Cancel)
                    {
                        _channel.Writer.TryComplete();
                        throw new OperationCanceledException(
                            item.CancelReason is null
                                ? "Request cancelled by runtime"
                                : $"Request cancelled by runtime: {item.CancelReason}");
                    }

                    if (item.End)
                    {
                        _channel.Writer.TryComplete();
                        yield break;
                    }

                    if (item.Chunk is { Length: > 0 })
                    {
                        yield return item.Chunk;
                    }
                }
            }
        }

        private struct Item
        {
            public byte[]? Chunk;
            public bool End;
            public bool Cancel;
            public string? CancelReason;
        }
    }

    private sealed class AdapterResponseSink(
        string requestId,
        PendingState state,
        Func<ILlmInferenceResponseChannel?> getChannel,
        ConcurrentDictionary<string, PendingState> pending) : LlmInferenceResponseSink
    {
        public override async Task StartAsync(LlmInferenceResponseInit init)
        {
            ArgumentNullException.ThrowIfNull(init);

            if (state.Started)
            {
                throw new InvalidOperationException("LLM inference response sink StartAsync() called twice.");
            }

            if (state.Finished)
            {
                throw new InvalidOperationException("LLM inference response sink already finished.");
            }

            state.Started = true;
            var result = await Channel()
                .HttpResponseStartAsync(requestId, init.Status, ToWireHeaders(init.Headers), init.StatusText)
                .ConfigureAwait(false);
            if (!result.Accepted)
            {
                RejectedByRuntime();
            }
        }

        public override Task WriteAsync(ReadOnlyMemory<byte> data) =>
            WriteChunk(Convert.ToBase64String(data.ToArray()), binary: true);

        public override Task WriteAsync(string text)
        {
            ArgumentNullException.ThrowIfNull(text);
            return WriteChunk(text, binary: false);
        }

        public override async Task EndAsync()
        {
            if (state.Finished)
            {
                return;
            }

            state.Finished = true;
            pending.TryRemove(requestId, out _);
            await Channel().HttpResponseChunkAsync(requestId, string.Empty, end: true).ConfigureAwait(false);
        }

        public override async Task ErrorAsync(string message, string? code = null)
        {
            ArgumentNullException.ThrowIfNull(message);

            if (state.Finished)
            {
                return;
            }

            state.Finished = true;
            pending.TryRemove(requestId, out _);
            await Channel()
                .HttpResponseChunkAsync(
                    requestId,
                    string.Empty,
                    end: true,
                    error: new LlmInferenceHttpResponseChunkError { Message = message, Code = code })
                .ConfigureAwait(false);
        }

        private async Task WriteChunk(string data, bool binary)
        {
            if (state.Cancelled)
            {
                throw new InvalidOperationException("LLM inference request was cancelled by the runtime.");
            }

            if (!state.Started)
            {
                throw new InvalidOperationException("LLM inference response sink WriteAsync() called before StartAsync().");
            }

            if (state.Finished)
            {
                throw new InvalidOperationException("LLM inference response sink WriteAsync() called after EndAsync()/ErrorAsync().");
            }

            var result = await Channel()
                .HttpResponseChunkAsync(requestId, data, binary: binary, end: false)
                .ConfigureAwait(false);
            if (!result.Accepted)
            {
                RejectedByRuntime();
            }
        }

        private ILlmInferenceResponseChannel Channel() =>
            getChannel() ?? throw new InvalidOperationException("LLM inference response sink used after RPC connection closed.");

        // The runtime acknowledges every response frame with accepted; accepted:
        // false means it has dropped the request (e.g. it cancelled), so we abort
        // the provider's upstream work and stop emitting.
        private void RejectedByRuntime()
        {
            if (!state.Cancelled)
            {
                state.Cancelled = true;
                state.Abort.Cancel();
            }

            state.Finished = true;
            pending.TryRemove(requestId, out _);
            throw new InvalidOperationException("LLM inference response was rejected by the runtime (request no longer active).");
        }

        private static Dictionary<string, IList<string>> ToWireHeaders(IReadOnlyDictionary<string, IReadOnlyList<string>>? headers)
        {
            var result = new Dictionary<string, IList<string>>(StringComparer.OrdinalIgnoreCase);
            if (headers is null)
            {
                return result;
            }

            foreach (var (name, values) in headers)
            {
                result[name] = values as IList<string> ?? [.. values];
            }

            return result;
        }
    }
}

/// <summary>
/// Minimal seam over the runtime-bound <c>llmInference</c> server API the
/// adapter uses to push response frames back to the runtime. Extracted as an
/// interface so the adapter can be unit-tested without a live JSON-RPC
/// connection.
/// </summary>
internal interface ILlmInferenceResponseChannel
{
    Task<LlmInferenceHttpResponseStartResult> HttpResponseStartAsync(string requestId, long status, IDictionary<string, IList<string>> headers, string? statusText = null);

    Task<LlmInferenceHttpResponseChunkResult> HttpResponseChunkAsync(string requestId, string data, bool? binary = null, bool? end = null, LlmInferenceHttpResponseChunkError? error = null);
}

/// <summary>
/// Production <see cref="ILlmInferenceResponseChannel"/> backed by the generated
/// <see cref="ServerRpc"/> client.
/// </summary>
internal sealed class ServerRpcResponseChannel(ServerRpc serverRpc) : ILlmInferenceResponseChannel
{
    public Task<LlmInferenceHttpResponseStartResult> HttpResponseStartAsync(string requestId, long status, IDictionary<string, IList<string>> headers, string? statusText = null) =>
        serverRpc.LlmInference.HttpResponseStartAsync(requestId, status, headers, statusText);

    public Task<LlmInferenceHttpResponseChunkResult> HttpResponseChunkAsync(string requestId, string data, bool? binary = null, bool? end = null, LlmInferenceHttpResponseChunkError? error = null) =>
        serverRpc.LlmInference.HttpResponseChunkAsync(requestId, data, binary, end, error);
}

File: python/copilot/llm_inference_provider.py
# --------------------------------------------------------------------------------------------
#  Copyright (c) Microsoft Corporation. All rights reserved.
# --------------------------------------------------------------------------------------------

"""Low-level LLM inference provider types and the RPC adapter.

The SDK consumer implements :class:`LlmInferenceProvider` (usually by
subclassing the idiomatic :class:`~copilot.llm_request_handler.LlmRequestHandler`).
:func:`create_llm_inference_adapter` converts a provider into an object that
conforms to the generated :class:`~copilot.generated.rpc.LlmInferenceHandler`
protocol, wiring the inbound ``httpRequestStart`` / ``httpRequestChunk`` frames
into the provider and translating the provider's response writes back into
outbound ``httpResponseStart`` / ``httpResponseChunk`` RPCs.
"""

from __future__ import annotations

import asyncio
import base64
from collections.abc import AsyncIterator, Awaitable, Callable
from dataclasses import dataclass, field
from typing import Protocol, runtime_checkable

from .generated.rpc import (
    LlmInferenceHTTPRequestChunkRequest,
    LlmInferenceHTTPRequestChunkResult,
    LlmInferenceHTTPRequestStartRequest,
    LlmInferenceHTTPRequestStartResult,
    LlmInferenceHTTPResponseChunkError,
    LlmInferenceHTTPResponseChunkRequest,
    LlmInferenceHTTPResponseStartRequest,
    ServerLlmInferenceApi,
)

# Headers are multi-valued: a header name maps to a list of values.
LlmInferenceHeaders = dict[str, list[str]]


@dataclass
class LlmInferenceResponseInit:
    """Response head passed to :meth:`LlmInferenceResponseSink.start`."""

    status: int
    status_text: str | None = None
    headers: LlmInferenceHeaders | None = None


@runtime_checkable
class LlmInferenceResponseSink(Protocol):
    """Sink the consumer writes the upstream response into.

    The state machine is strict: ``start`` once, then zero or more ``write``
    calls, finishing with exactly one of ``end`` or ``error``. Calling out of
    order raises.
    """

    async def start(self, init: LlmInferenceResponseInit) -> None:
        """Send the response head (status + headers) back to the runtime."""
        ...

    async def write(self, data: str | bytes) -> None:
        """Send a body chunk. ``str`` is encoded as UTF-8; ``bytes`` is sent as binary."""
        ...

    async def end(self) -> None:
        """Mark end-of-stream cleanly."""
        ...

    async def error(self, message: str, code: str | None = None) -> None:
        """Mark end-of-stream with a transport-level failure."""
        ...


@dataclass
class LlmInferenceRequest:
    """An outbound model-layer HTTP request the runtime is asking the SDK to handle.

    This is a low-level shape: URL / method / headers verbatim, body bytes
    delivered as an async iterator, response delivered through
    :attr:`response_body`. The runtime does not classify the request; consumers
    that need a provider type or endpoint kind derive it from the URL / headers.
    """

    request_id: str
    """Opaque runtime-minted id, stable across the request lifecycle."""

    method: str
    """HTTP method (``GET``, ``POST``, ...)."""

    url: str
    """Absolute URL."""

    headers: LlmInferenceHeaders
    """HTTP request headers, multi-valued."""

    transport: str
    """``"http"`` (plain HTTP / SSE) or ``"websocket"`` (full-duplex channel)."""

    request_body: AsyncIterator[bytes]
    """Request body bytes, yielded as they arrive. Empty bodies yield zero chunks."""

    cancel_event: asyncio.Event
    """Set when the runtime cancels this in-flight request. Pass it through to
    your transport so the upstream call is torn down too. After it fires, writes
    to :attr:`response_body` are ignored."""

    response_body: LlmInferenceResponseSink
    """Sink the consumer writes the upstream response into."""

    session_id: str | None = None
    """Id of the runtime session that triggered this request, when in scope.
    Absent for out-of-session requests (e.g. the startup model catalog)."""


@runtime_checkable
class LlmInferenceProvider(Protocol):
    """Interface for an LLM inference provider.

    The consumer implements :meth:`on_llm_request`. The same callback handles
    both buffered and streaming responses; the consumer just calls
    ``response_body.write`` zero or more times before ``end``.
    """

    async def on_llm_request(self, request: LlmInferenceRequest) -> None:
        """Service a single outbound LLM HTTP request.

        The consumer must eventually call either ``response_body.end()`` or
        ``response_body.error(...)``; failing to do so leaks runtime state.
        Raising surfaces a transport-level failure to the runtime.
        """
        ...


@dataclass
class LlmInferenceConfig:
    """Connection-level LLM inference callback configuration.

    Passed as the ``llm_inference`` client option. The ``handler`` is registered
    process-wide and invoked for every model-layer HTTP/WebSocket request the
    runtime would otherwise issue, for both BYOK and CAPI traffic.
    """

    handler: LlmInferenceProvider



@dataclass
class _BodyItem:
    chunk: bytes | None = None
    end: bool = False
    cancel: bool = False
    cancel_reason: str | None = None


class _BodyQueue:
    """An async iterator of request-body byte chunks fed by the runtime."""

    def __init__(self) -> None:
        self._queue: asyncio.Queue[_BodyItem] = asyncio.Queue()
        self._done = False

    def push(self, item: _BodyItem) -> None:
        self._queue.put_nowait(item)

    def __aiter__(self) -> AsyncIterator[bytes]:
        return self

    async def __anext__(self) -> bytes:
        if self._done:
            raise StopAsyncIteration
        item = await self._queue.get()
        if item.cancel:
            self._done = True
            reason = (
                f"Request cancelled by runtime: {item.cancel_reason}"
                if item.cancel_reason
                else "Request cancelled by runtime"
            )
            raise RuntimeError(reason)
        if item.end:
            self._done = True
            raise StopAsyncIteration
        return item.chunk if item.chunk is not None else b""


@dataclass
class _PendingState:
    queue: _BodyQueue
    cancel_event: asyncio.Event
    started: bool = False
    finished: bool = False
    cancelled: bool = False
    task: asyncio.Task[None] | None = field(default=None)


def _decode_chunk_data(data: str, binary: bool) -> bytes:
    if binary:
        return base64.b64decode(data)
    return data.encode("utf-8")


class _RuntimeRejectedError(RuntimeError):
    """Raised when the runtime drops an in-flight request (``accepted: False``)."""


def create_llm_inference_adapter(
    provider: LlmInferenceProvider,
    get_server_rpc: Callable[[], ServerLlmInferenceApi | None],
) -> "_LlmInferenceAdapter":
    """Adapt an :class:`LlmInferenceProvider` into the generated handler shape.

    Maintains a per-``request_id`` state table: each ``http_request_start``
    allocates a body queue + response sink and fires ``provider.on_llm_request``
    in the background. Subsequent ``http_request_chunk`` frames are routed into
    the queue. The sink translates ``start`` / ``write`` / ``end`` / ``error``
    calls into outbound ``httpResponseStart`` / ``httpResponseChunk`` RPCs.

    ``http_request_start`` returns immediately after registering state so the
    runtime's RPC reply is not gated on the consumer's I/O.
    """
    return _LlmInferenceAdapter(provider, get_server_rpc)


class _LlmInferenceAdapter:
    def __init__(
        self,
        provider: LlmInferenceProvider,
        get_server_rpc: Callable[[], ServerLlmInferenceApi | None],
    ) -> None:
        self._provider = provider
        self._get_server_rpc = get_server_rpc
        self._pending: dict[str, _PendingState] = {}
        # Defense-in-depth backstop: chunks that arrive before their start frame
        # (a reordering the runtime's single ordered dispatch should make
        # impossible) are staged here and drained the moment the matching
        # http_request_start registers state, so a body byte is never dropped.
        self._staged: dict[str, list[LlmInferenceHTTPRequestChunkRequest]] = {}

    def _route_chunk(self, state: _PendingState, params: LlmInferenceHTTPRequestChunkRequest) -> None:
        if params.cancel:
            state.cancelled = True
            state.cancel_event.set()
            state.queue.push(_BodyItem(cancel=True, cancel_reason=params.cancel_reason))
            return
        if params.data:
            state.queue.push(_BodyItem(chunk=_decode_chunk_data(params.data, bool(params.binary))))
        if params.end:
            state.queue.push(_BodyItem(end=True))

    def _require_rpc(self) -> ServerLlmInferenceApi:
        rpc = self._get_server_rpc()
        if rpc is None:
            raise RuntimeError("LLM inference response sink used after RPC connection closed.")
        return rpc

    def _make_sink(self, request_id: str, state: _PendingState) -> LlmInferenceResponseSink:
        adapter = self

        def reject() -> None:
            # The runtime acknowledges every response frame with ``accepted``.
            # ``accepted: False`` means it has dropped the request, so we abort
            # the provider's upstream work and stop emitting.
            if not state.cancelled:
                state.cancelled = True
                state.cancel_event.set()
            state.finished = True
            adapter._pending.pop(request_id, None)
            raise _RuntimeRejectedError(
                "LLM inference response was rejected by the runtime (request no longer active)."
            )

        class _Sink:
            async def start(self, init: LlmInferenceResponseInit) -> None:
                if state.started:
                    raise RuntimeError("LLM inference response sink.start() called twice.")
                if state.finished:
                    raise RuntimeError("LLM inference response sink already finished.")
                state.started = True
                result = await adapter._require_rpc().http_response_start(
                    LlmInferenceHTTPResponseStartRequest(
                        headers=init.headers or {},
                        request_id=request_id,
                        status=init.status,
                        status_text=init.status_text,
                    )
                )
                if not result.accepted:
                    reject()

            async def write(self, data: str | bytes) -> None:
                if state.cancelled:
                    raise RuntimeError("LLM inference request was cancelled by the runtime.")
                if not state.started:
                    raise RuntimeError("LLM inference response sink.write() called before start().")
                if state.finished:
                    raise RuntimeError("LLM inference response sink.write() called after end()/error().")
                is_binary = isinstance(data, bytes | bytearray)
                payload = (
                    base64.b64encode(bytes(data)).decode("ascii")
                    if is_binary
                    else str(data)
                )
                result = await adapter._require_rpc().http_response_chunk(
                    LlmInferenceHTTPResponseChunkRequest(
                        data=payload,
                        request_id=request_id,
                        binary=is_binary or None,
                        end=False,
                    )
                )
                if not result.accepted:
                    reject()

            async def end(self) -> None:
                if state.finished:
                    return
                state.finished = True
                adapter._pending.pop(request_id, None)
                await adapter._require_rpc().http_response_chunk(
                    LlmInferenceHTTPResponseChunkRequest(data="", request_id=request_id, end=True)
                )

            async def error(self, message: str, code: str | None = None) -> None:
                if state.finished:
                    return
                state.finished = True
                adapter._pending.pop(request_id, None)
                await adapter._require_rpc().http_response_chunk(
                    LlmInferenceHTTPResponseChunkRequest(
                        data="",
                        request_id=request_id,
                        end=True,
                        error=LlmInferenceHTTPResponseChunkError(message=message, code=code),
                    )
                )

        return _Sink()

    async def _fail_via_sink(
        self, sink: LlmInferenceResponseSink, state: _PendingState, message: str
    ) -> None:
        if state.finished:
            return
        try:
            if not state.started:
                await sink.start(LlmInferenceResponseInit(status=502))
            await sink.error(message)
        except Exception:
            # Best-effort — the connection may already be dead.
            pass

    async def _finish_cancelled(self, sink: LlmInferenceResponseSink, state: _PendingState) -> None:
        if state.finished:
            return
        try:
            if not state.started:
                await sink.start(LlmInferenceResponseInit(status=499))
            await sink.error("Request cancelled by runtime", code="cancelled")
        except Exception:
            # Best-effort — the runtime already dropped the request on cancel.
            pass

    async def _run_provider(
        self, request: LlmInferenceRequest, sink: LlmInferenceResponseSink, state: _PendingState
    ) -> None:
        try:
            await self._provider.on_llm_request(request)
            if not state.finished:
                await self._fail_via_sink(
                    sink,
                    state,
                    "LLM inference provider returned without finalising the response "
                    "(call response_body.end() or .error()).",
                )
        except _RuntimeRejectedError:
            # The runtime already dropped the request; nothing more to emit.
            pass
        except Exception as exc:
            if state.cancelled or state.cancel_event.is_set():
                await self._finish_cancelled(sink, state)
                return
            await self._fail_via_sink(sink, state, str(exc))

    async def http_request_start(
        self, params: LlmInferenceHTTPRequestStartRequest
    ) -> LlmInferenceHTTPRequestStartResult:
        state = _PendingState(queue=_BodyQueue(), cancel_event=asyncio.Event())
        self._pending[params.request_id] = state

        staged = self._staged.pop(params.request_id, None)
        if staged:
            for chunk in staged:
                self._route_chunk(state, chunk)

        sink = self._make_sink(params.request_id, state)
        transport = (
            params.transport.value if params.transport is not None else "http"
        )
        request = LlmInferenceRequest(
            request_id=params.request_id,
            session_id=params.session_id,
            method=params.method,
            url=params.url,
            headers=params.headers,
            transport=transport,
            request_body=state.queue,
            cancel_event=state.cancel_event,
            response_body=sink,
        )
        state.task = asyncio.create_task(self._run_provider(request, sink, state))
        return LlmInferenceHTTPRequestStartResult()

    async def http_request_chunk(
        self, params: LlmInferenceHTTPRequestChunkRequest
    ) -> LlmInferenceHTTPRequestChunkResult:
        state = self._pending.get(params.request_id)
        if state is None:
            self._staged.setdefault(params.request_id, []).append(params)
            return LlmInferenceHTTPRequestChunkResult()
        self._route_chunk(state, params)
        return LlmInferenceHTTPRequestChunkResult()
